solc/CommandLineParser.cpp

```diff
@@ -840,13 +840,20 @@ bool CommandLineParser::processArgs()
 	else
 		m_options.input.mode = InputMode::Compiler;
 
-	if (
-		m_args.count(g_strExperimentalViaIR) > 0 &&
-		m_options.input.mode != InputMode::Compiler &&
-		m_options.input.mode != InputMode::CompilerWithASTImport
-	)
+	map<string, set<InputMode>> validOptionInputModeCombinations = {
+		{g_strErrorRecovery, {InputMode::Compiler, InputMode::CompilerWithASTImport}},
+		{g_strExperimentalViaIR, {InputMode::Compiler, InputMode::CompilerWithASTImport}},
+	};
+	string invalidOptions;
+	for (auto const& [optionName, inputModes]: validOptionInputModeCombinations)
+	{
+		if (m_args.count(optionName) > 0 && inputModes.count(m_options.input.mode) == 0)
+			invalidOptions.append("--" + optionName + " ");
+	}
+
+	if (!invalidOptions.empty())
 	{
-		serr() << "The option --" << g_strExperimentalViaIR << " is only supported in the compiler mode." << endl;
+		serr() << "The option " << invalidOptions << "is not supported in the current mode." << endl;
```

Comment:
This message only works in case there's one invalid option.
```suggestion
		serr() << "The following options are not supported in the current input mode: " << invalidOptions << endl;
```